src/GatewayProvider.ts

```diff
@@ -0,0 +1,80 @@
+import { Chain } from './types.js';
+import { withResolvers } from './utils.js';
+import {
+  JsonRpcPayload,
+  JsonRpcProvider,
+  JsonRpcResult,
+} from 'ethers/providers';
+import { FetchRequest } from 'ethers/utils';
+
+function shouldNeverBatch(payload: JsonRpcPayload) {
+  // see: LineaProver.ts:fetchProofs()
+  return payload.method === 'linea_getProof';
+}
+
+export class GatewayProvider extends JsonRpcProvider {
+  constructor(
+    fr: FetchRequest,
+    readonly chain: Chain
+  ) {
+    super(fr, chain, { staticNetwork: true });
+  }
+  override async _send(
+    payload: JsonRpcPayload | JsonRpcPayload[]
+  ): Promise<JsonRpcResult[]> {
+    if (Array.isArray(payload)) {
+      const ps: Promise<any>[] = [];
+      const batch: JsonRpcPayload[] = [];
+      const { promise, resolve, reject } = withResolvers<JsonRpcResult[]>();
+      const ret = payload.map((x) => {
+        if (shouldNeverBatch(x)) {
+          const p = this._sendWithRetry(x).then((x) => x[0]);
+          ps.push(p);
+          return p;
+        } else {
+          const i = batch.length;
+          batch.push(x);
+          return promise.then((v) => v[i]);
+        }
+      });
+      if (batch.length) {
+        ps.push(this._sendWithRetry(batch).then(resolve, reject));
+        await Promise.all(ps);
+      }
+      return Promise.all(ret);
+    }
+    return this._sendWithRetry(payload);
+  }
+  async _sendWithRetry(
+    payload: JsonRpcPayload | JsonRpcPayload[]
+  ): Promise<JsonRpcResult[]> {
+    let backoff = 250;
+    for (let attempt = 0; ; attempt++) {
+      try {
+        // ethers bug: this type is wrong
```

Comment:
which type? do you have a link ?

Reply:
Send returns `(JsonRpcResult | JsonRpcErrror)[]` not `JsonRpcResult[]`